File: FloodIt.java
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Scanner;
import java.util.Map.Entry;

/**
 * FloodIt
 */
public class FloodIt {

    private static int[][] grid;
    private static List<Entry<Integer, Integer>> flooded = new ArrayList<Entry<Integer, Integer>>() {
        {
            Map.entry(1, 1);
        }
    };
    private static int moves = 0;

    private static void printGrid() {
        for (int i = 1; i < grid.length - 1; i++) {
            for (int j = 1; j < grid.length - 1; j++) {
                int elem = grid[i][j];
                if (flooded.contains(Map.entry(i, j))) {
                    Output.print("███", Output.Color.fromOrdinal(elem));
                } else {
                    Output.print(" " + Integer.toString(elem) + " ", Output.Color.fromOrdinal(elem));
                }
            }
            System.out.println();
        }
    }

    private static int[][] createGrid(int size) {
        int[][] tempGrid = new int[size + 2][size + 2];

        for (int i = 0; i <= size + 1; i++) {
            for (int j = 0; j <= size + 1; j++) {
                if (i == 0 || i == size + 1 || j == 0 || j == size + 1) {
                    tempGrid[i][j] = -1;
                } else {
                    tempGrid[i][j] = (int) Math.floor(Math.random() * 6);
                }
            }
        }
        return tempGrid;
    }

    private static void traverse(int x, int y, int move) {
        for (int[] offset : new int[][] { { -1, 0 }, { 0, -1 }, { 0, 1 }, { 1, 0 } }) {
            int i = x + offset[0];
            int j = y + offset[1];
            if (grid[i][j] == move && !flooded.contains(Map.entry(i, j))) {
                flooded.add(Map.entry(i, j));
                traverse(i, j, move);
            }
        }
    }

    private static void playMove(int move) {

        for (Entry<Integer, Integer> coord : new ArrayList<>(flooded)) {
            int x = coord.getKey(), y = coord.getValue();
            // Change the flooded bit
            grid[x][y] = move;
            // And traverse it and its neighbours recursively
            traverse(x, y, move);
        }

    }

    private static void initialise(int n) {
        grid = createGrid(n);
        moves = 0;
        flooded.clear();
        flooded.add(Map.entry(1, 1));

        traverse(1, 1, grid[1][1]);
        Output.clearScreen();
    }

    private static int parseMoveFromInput(String str) {
        try {
            int nextMove = Integer.parseInt(str);
            if (nextMove > 6 || grid[1][1] == nextMove) {
                return -1;
            }
            return nextMove;
        } catch (Exception e) {
            return -1;
        }

    }

    public static void main(String[] args) {
        int n = Integer.parseInt(args[0]);
        initialise(n);

        Scanner sc = new Scanner(System.in);

        while (true) {
            Output.clearScreen();
            printGrid();
            if (flooded.size() >= (n * n)) {
                break;
            }

            while (true) {
                System.out.println("Moves: " + moves + " | Flooded: " + flooded.size());
                System.out.print("Enter move: ");

                String str = sc.nextLine();
                int nextMove = parseMoveFromInput(str);

                if (nextMove != -1) {
                    moves++;
                    playMove(nextMove);
                    break;
                }
                Output.clearLastLines(2);
            }
        }
        // Output.clearLastLines(1);
        System.out.println();
        System.out.println("You won in " + Integer.toString(moves) + " move(s).");
        System.out.print("Enter 's' to start again or any other key to quit. ");
        String cmd = sc.nextLine();
        if (cmd.equals("s")) {
            main(args);
        }
        sc.reset();
        sc.close();
    }

}

File: go/floodit.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"strings"

	"github.com/muesli/termenv"
)

type gridItem struct {
	value   int
	flooded bool
}

var grid [][]gridItem
var nFlooded int
var moves int
var p = termenv.ColorProfile()

var colors = []string{
	"#E88388", //red
	"#A8CC8C", //green
	"#DBAB79", //yellow
	"#71BEF2", //blue
	"#D290E4", //magenta
	"#FFFFF",  //gray
}
var bg = termenv.BackgroundColor()

func printGrid() {
	for i := 1; i < len(grid)-1; i++ {
		for j := 1; j < len(grid)-1; j++ {
			elem := grid[i][j]
			elemStr := strconv.Itoa(elem.value)
			out := termenv.String(" " + elemStr + " ")

			if elem.flooded {
				out = out.Foreground(bg).Background(p.Color(colors[elem.value]))
			} else {
				out = out.Foreground(p.Color(colors[elem.value]))
			}
			fmt.Print(out)
		}
		fmt.Println()
	}
	fmt.Println()
}

func initialiseGrid(size int) [][]gridItem {
	tempGrid := make([][]gridItem, size+2)

	for i := 0; i <= size+1; i++ {
		tempGrid[i] = make([]gridItem, size+2)
		for j := 0; j <= size+1; j++ {
			if i == 0 || i == size+1 || j == 0 || j == size+1 {
				tempGrid[i][j] = gridItem{-1, false}
			} else {
				tempGrid[i][j] = gridItem{int(math.Floor(rand.Float64() * 6)), false}
			}
		}
	}
	return tempGrid
}

func flood(x, y, move int) {
	offsets := [][]int{{-1, 0}, {0, -1}, {0, 1}, {1, 0}}
	for _, offset := range offsets {
		i := x + offset[0]
		j := y + offset[1]
		elem := grid[i][j]
		if elem.value == move && !elem.flooded {
			grid[i][j].flooded = true
			nFlooded++
			flood(i, j, move)
		}
	}
}

func playMove(move int) {

	for x, row := range grid {
		for y, elem := range row {
			if elem.flooded {
				// Change the flooded bit
				grid[x][y].value = move
				// And flood it and its neighbours recursively
				flood(x, y, move)
			}
		}
	}
}

func initialise(n int) {
	grid = initialiseGrid(n)
	grid[1][1].flooded = true
	nFlooded = 1
	moves = 0
	flood(1, 1, grid[1][1].value)
}

func parseMoveFromInput(str string) int {
	if strings.TrimSpace(str) == "" {
		return -1
	}

	nextMove, err := strconv.Atoi(str)

	if err != nil {
		return -1
	}
	if nextMove > 6 || grid[1][1].value == nextMove {
		return -1
	}
	return nextMove
}

func parseGridSizeFromArgs(args []string) int {
	if len(args) <= 1 {
		return 5
	}
	parsed, err := strconv.Atoi(args[1])
	if err != nil {
		return 5
	}
	size := int(math.Min(math.Max(5, float64(parsed)), 21))
	return size
}

func main() {
	{
		size := parseGridSizeFromArgs(os.Args)
		initialise(size)

		for {

			termenv.ClearScreen()
			printGrid()
			if nFlooded >= (size * size) {
				break
			}

			for {
				fmt.Println("Moves: " + strconv.Itoa(moves) + " | Flooded: " + strconv.Itoa(nFlooded))
				fmt.Print("Enter move: ")

				str := ""
				fmt.Scanln(&str)
				nextMove := parseMoveFromInput(str)

				if nextMove != -1 {
					moves++
					playMove(nextMove)
					break
				}
				termenv.ClearLines(2)
			}
		}

		fmt.Println("You won in " + strconv.Itoa(moves) + " move(s).")
		fmt.Print("Enter 's' to start again or any other key to quit. ")
		cmd := ""
		fmt.Scanln(&cmd)
		if cmd == "s" {
			main()
		}
	}
}
